### ISSUE TITLE
removal of `Unpin` requirement on `next_poll`

### ISSUE THREAD

**Issue by jjl** [irrustible/futures-micro#9]
I cannot for the life of me figure out a way to do this without either using unsafe or turning it into a struct like futures-lite does.

I had thought we'd be able to do something with `pin!()`, but getting it back out of the pin is problematic...

**Comment by fogti** [comment]
I don't think that's easily possible. Maybe the abstraction, as currently present, isn't optimal. It might be a good idea to implement something which fits (your use-case) independent of `next_poll`, maybe as a raw `Future`-implementing struct. The naive way of just removing the `Unpin` requirement and using `unsafe` in the implementation is just not always sound afaik. Maybe we could add an `unsafe` version which doesn't has the `Unpin` requirement...

**Comment by jjl** [comment]
I don't think I actually use it anywhere. I believe it's actually quite easy to implement without compromising soundness with a sprinkling of unsafe, but then we're back at needing to trust my unsafe code.

WIP:

```rust
diff --git a/src/lib.rs b/src/lib.rs
index 6e2f98d..de00e0f 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -251,23 +251,18 @@ pub fn sleep() -> impl Future<Output = ()> {
 /// assert_eq!(f.await, 1);
 /// # })
 /// ```
-pub fn next_poll<F>(f: F) -> impl Future<Output = Result<F::Output, F>>
+pub fn next_poll<'a, F: 'a>(f: F) -> impl Future<Output = Result<F::Output, F>> + 'a
 where
-    F: Future + Unpin,
+    F: Future
 {
-    let mut f = Some(f);
+    let g = Some(f);
+    pin!(g);
     poll_fn(move |ctx| {
-        Poll::Ready(
-            match Pin::new(
-                f.as_mut()
-                    .expect("`next_poll` Future polled after completion"),
-            )
-            .poll(ctx)
-            {
-                Poll::Ready(val) => Ok(val),
-                Poll::Pending => Err(f.take().unwrap()),
-            },
-        )
+        let h = g.as_pin_mut().expect("next_poll: Future polled after completion!");
+        match F::poll(h, ctx) {
+            Poll::Ready(val) => Poll::Ready(Ok(val)),
+            Poll::Pending => Poll::Ready(Err(g.as_mut().take().unwrap()))
+        }
     })
 }
```

**Comment by fogti** [comment]
I don't think you can move a future after it was pinned. An alternative might be to take the future by mutable reference.
e.g.:
```rust
pub fn next_poll<F>(f: &mut Pin<F>) -> impl Future<Output = Result<F::Output, ()>>
```

**Comment by jjl** [comment]
ah yes, if we pull the pinning out to the caller it should work.

on the other hand,  dealing directly with the pinning as a user isn't very ergonomic. not having to think about it makes async/await syntax somewhat easier to use.

**Comment by jjl** [comment]
a macro would be an option, it would just use `pin!`, but ugh, more macros

**Comment by fogti** [comment]
The thing is: Is this even a useful abstraction? If it is too limited in its current form, because of the requirement of `Unpin`, but the version where the caller does the pinning is also unergonomic, maybe we should drop this abstraction until we come up with a better one. It does not offer much over "just implementing `Future` yourself".

**Comment by jjl** [comment]
as i said before, i don't actually think i use it anywere, and while i can see potential use cases, i'm not sure it's the same level of utility as the other select list of things we choose to include.

i can't speak for other people, but unless i'm working in no_std, i generally pull in futures-lite too and this is in there along with a much larger set of batteries. futures-lite does not have the variadic `or!` and `zip!` macros last i checked though, so it's not like it obviates the need to pull in `futures-micro`.

and with that, while i totally understand stjepang wanting to spend less time maintaining rust software, the whole situation with futures-lite is a bit weird now and i'm not sure if it will ultimately depend on us. i'm not sure i'm particularly bothered either, i wrote this crate for fun and it seems people have had some good use out of it so *shrug* we can go our own way if that's what we're supposed to do.

**Comment by jjl** [comment]
Closed by 1b6218a 

Anything else you'd like to see before a new release?

**Comment by fogti** [comment]
No, I have no objections to this.
